### PR TITLE
[MC] Emit a jump table size section

### PR DESCRIPTION
**Pull request by omern1**
This patch will make LLVM emit a jump table size section containing tuples of (jump table address, entry count) in object files. This section is useful for tools that need to statically reconstruct the control flow of executables.

The name of the new section is .debug_llvm_jump_table_sizes because that makes both llvm-strip and GNU strip remove it.

At the moment this is only enabled by default for the PS5 target.

### CODE REVIEW

**Comment by compnerd** [comment]
The `.debug_` prefix is for DEARF content, is this being proposed to the DWARF standard?

**Comment by omern1** [comment]
> The `.debug_` prefix is for DEARF content, is this being proposed to the DWARF standard?

No, this isn't being proposed for the DWARF standard. Though, searching through the DWARF standard I couldn't find anywhere saying that the `.debug_` section name prefix is reserved for DWARF information. Similarly the SysV ABI and the ELF spec don't seem to say anything like that either. I suspect this might be a namespacing convention rather than a requirement.

I feel like there's a case for this prefix as GNU `strip` doesn't seem to strip sections with the `.llvm_` prefix and we don't people to accidentally ship binaries with this section in them.

Also, the choice of `.debug_llvm_` is specifically to avoid clashes with DWARF sections.

**Comment by MaskRay** [comment]
Conventionally, `.debug_*` sections are reserved for DWARF, even without official standards.
GNU strip's default behavior --strip-all, emulated by llvm-objcopy's --strip-all-gnu option, is roughly to remove non-SHF_ALLOC sections that are not debug.
The `.llvm` prefix holds no special meaning.

When designing new sections, we should not adapt to the --strip-all-gnu behavior.

Jump tables exhibit diverse structures. For instance, consider llvm/test/CodeGen/ARM/jump-table-tbh.ll and Mach-O's .data_region.

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 1]
```diff
@@ -2764,6 +2770,27 @@ void AsmPrinter::emitJumpTableInfo() {
     for (const MachineBasicBlock *MBB : JTBBs)
       emitJumpTableEntry(MJTI, MBB, JTI);
   }
+
+  if (EmitJumpTableSizesSection && TM.getTargetTriple().isOSBinFormatELF() &&
```
Comment (compnerd) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:2774]: Outside of flags and section name, this is pretty generic. Why not open this to all object fully formats?

File: llvm/test/CodeGen/X86/jump-table-size-section.ll [hunk 2]
```diff
@@ -0,0 +1,97 @@
+; RUN: llc %s -o - -emit-jump-table-sizes-section -verify-machineinstrs --relocation-model=pic | FileCheck --check-prefix=CHECK %s
+; RUN: llc %s -o - -verify-machineinstrs --relocation-model=pic | FileCheck --check-prefix=NOFLAG %s
```
Comment (compnerd) [inline llvm/test/CodeGen/X86/jump-table-size-section.ll:2]: You will need explicit triples - these tests may be run where the default target is not PS5 and the default object file format is not ELF.

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp (hunk 1 above)
Comment (omern1) [reply llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp]: I've added support for COFF (hopefully correctly). I'm completely unfamiliar with MachO so I suggest that we leave that for a followup.

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 3]
```diff
@@ -2764,10 +2771,59 @@ void AsmPrinter::emitJumpTableInfo() {
...
+  } else if (TM.getTargetTriple().isOSBinFormatCOFF()) {
+    if (F.hasComdat()) {
+      MCSymbolCOFF *LinkedToSym = dyn_cast<MCSymbolCOFF>(CurrentFnSym);
+
+      JumpTableSizesSection = OutContext.getCOFFSection(
+          sectionName,
+          COFF::IMAGE_SCN_CNT_INITIALIZED_DATA | COFF::IMAGE_SCN_MEM_READ |
+              COFF::IMAGE_SCN_LNK_COMDAT | COFF::IMAGE_SCN_MEM_DISCARDABLE,
+          F.getComdat()->getName(), COFF::IMAGE_COMDAT_SELECT_ASSOCIATIVE);
+    } else
```
Comment (compnerd) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:2809]: Braces here would be nice to match the other side.

File: llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp [hunk 4]
```diff
@@ -2764,10 +2771,59 @@ void AsmPrinter::emitJumpTableInfo() {
...
+      JumpTableSizesSection = OutContext.getCOFFSection(
+          sectionName, COFF::IMAGE_SCN_CNT_INITIALIZED_DATA |
+                           COFF::IMAGE_SCN_MEM_READ |
+                           COFF::IMAGE_SCN_MEM_DISCARDABLE);
+  } else {
+    return;
```
Comment (compnerd) [inline llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:2815]: Can you change this to an early return please?

**Review by compnerd — APPROVED** [review]

**Comment by ruotongyu** [comment]
Hi, I really like this design and am trying to apply it to more architectures, such as x86. However, it seems that the jump table is not properly mapped to the instruction that references it. Do you have any suggestions on how to correctly associate the jump table with the corresponding assembly instruction? Please let me know if I’m misunderstanding anything.

**Comment by omern1** [comment]
Hi @ruotongyu,

> and am trying to apply it to more architectures, such as x86.

I'm not sure what you mean here, this patch is really architecture independent and already works on X86. I maybe be misunderstanding you here though.

> Do you have any suggestions on how to correctly associate the jump table with the corresponding assembly instruction? Please let me know if I’m misunderstanding anything.

That's unfortunately something that I did not have time to do in this patch but https://github.com/llvm/llvm-project/pull/112606 is trying to achieve that.

**Pull request merged**